processing/src/main/java/org/apache/druid/segment/ColumnSelectorFactory.java

```diff
@@ -41,6 +43,19 @@ public interface ColumnSelectorFactory extends ColumnInspector
    */
   ColumnValueSelector makeColumnValueSelector(String columnName);
 
+  @Override
+  default int getColumnValueCardinality(String columnName)
+  {
+    final ColumnCapabilities capabilities = getColumnCapabilities(columnName);
+    if (capabilities == null) {
+      return DimensionDictionarySelector.CARDINALITY_UNKNOWN;
+    }
+    if (capabilities.is(ValueType.STRING) && capabilities.isDictionaryEncoded().isTrue()) {
+      final DimensionSelector dimensionSelector = makeDimensionSelector(DefaultDimensionSpec.of(columnName));
```

Comment:
Hmm I worry about performance here. The `QueryableIndex` implementation of `makeDimensionSelector` caches selectors, but most others don't.

I wonder if we can remove this method, so we don't have to worry about performance. There's only two call sites:

- One in `GroupingEngine`, where by the time this method is called, dimension selectors have already been created. The `GroupingEngine` could check the dimension selectors directly.
- One in `TopNQueryEngine`, where this method is called before the dimension selector is created, but it probably could be restructured...

Reply:
ok did this, I made a `GroupingSelector` interface for `GroupByColumnSelectorPlus` and `GroupByVectorColumnSelector` to implement for grouping engine cardinality computation, and reworked topn to make its selectorplus outside of the run method (which is nicer too since it doesnt make a new one for every granularity bucket, which was unnecessary after the previous refactor)